### PR TITLE
feat(adr-013): memory-store cutover preflight scanner

### LINKED ISSUE #5751: ADR-013 #1: memory-store cutover scanner (warn on missing rows)
## What to build

ADR-013 ([docs/dev/ADR-013-memory-store-consolidation.md](../blob/main/docs/dev/ADR-013-memory-store-consolidation.md)) Phase 6 is the cutover that retires the `decisions` table and stops dual-write to KNOWLEDGE.md. That step is destructive and one-way.

This slice ships the **non-destructive preflight**: on boot (or via an explicit doctor check), scan KNOWLEDGE.md and the `decisions` table for rows that have no corresponding entry in the `memories` table. Emit a workflow-logger warning + a persistent notification listing the count and a sample of unmigrated rows. No mutations, no schema changes — informational only.

Lets users see the gap before #6 (the cutover) lands, so the destructive step has zero surprise.

## Acceptance criteria

- [ ] A scanner function reads KNOWLEDGE.md rows + `decisions` rows and returns the set with no matching `memories` row (matched by content hash or scope+choice tuple).
- [ ] Scanner runs on session start (cheap — short-circuit when both `KNOWLEDGE.md` and `decisions` are empty or fully migrated).
- [ ] On detected gaps, emit (1) a workflow-logger warning, (2) a persistent notification with severity `warning`, message format: "Memory consolidation: N decisions / M KNOWLEDGE.md rows not yet in memories table — run /doctor for details."
- [ ] No mutations to any table or file.
- [ ] Test covers: zero-gap case (no warning), partial-gap case (correct count), all-stale case (correct count).

## Blocked by

None — can start immediately.

### PR DESCRIPTION
**Pull request by jeremymcs**
## Summary

Closes #5751. Unblocks #5755 (Phase 6 destructive cutover) once gaps read zero.

ADR-013 Phase 6 retires the `decisions` table and stops dual-write to `KNOWLEDGE.md`. The destructive step needs a way for users to see whether all rows have been absorbed into the `memories` table *before* anything is dropped. The existing `backfillDecisionsToMemories` (Phase 5) runs opportunistically on session start but reports nothing when work is unfinished.

This PR adds the preflight scanner. **Read-only — no DB writes, no file writes.**

### Two functions

- **`scanConsolidationGaps(basePath)`** — pure detection, returns a typed report (decisions surface + KNOWLEDGE.md surface + totals + a one-line summary).
- **`reportConsolidationGaps(basePath)`** — wraps the scan with a `workflow-logger` warning + persistent notification when gaps exist. Best-effort: never throws. Notification dedup (30s window) prevents repeat-boot flooding.

### Detection logic

**Decisions:** cross-checks active rows (`superseded_by IS NULL`) against `memories.structured_fields.sourceDecisionId` markers set by the existing `backfillDecisionsToMemories`. Consistent with the backfill's idempotency check pattern (LIKE on the JSON shape, anchored by quote characters to avoid prefix collisions like D1 vs D10).

**KNOWLEDGE.md:** parses the three legacy tables (Rules, Patterns, Lessons), cross-checks rows against a `sourceKnowledgeId` marker. Phase 6 will introduce that marker as part of the KNOWLEDGE.md → memories backfill. Until that path ships, every KNOWLEDGE.md row is honestly reported as unmigrated — which is the correct preflight state.

### Wiring

Called from `bootstrap/system-context.ts` immediately after `backfillDecisionsToMemories`. Fires on every session start; emits a notification only when gaps exist.

## Test plan

- [x] 10 tests in `memory-consolidation-scanner.test.ts` — parse edge cases, clean state, decisions-only gap, superseded-skip, KNOWLEDGE.md-only gap, combined gap, notification emit, silent-when-clean
- [x] Combined regression run with `memory-store.test.ts` (15 tests), `load-memory-block.test.ts` (16 tests), `notification-store.test.ts` (6 tests) — 47/47 pass

## Files

- `src/resources/extensions/gsd/memory-consolidation-scanner.ts` *(new, ~200 lines)*
- `src/resources/extensions/gsd/bootstrap/system-context.ts` — wiring (10 lines)
- `src/resources/extensions/gsd/tests/memory-consolidation-scanner.test.ts` *(new, ~280 lines, 10 tests)*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a startup preflight scanner that detects gaps between decisions/legacy knowledge and migrated memories, emitting a warning and recommending inspection.

* **Tests**
  * Added end-to-end tests covering parsing, gap scanning, and warning/reporting behavior for the consolidation scanner.

* **Documentation**
  * Updated ADR and troubleshooting guides with Phase‑6 preflight behavior and operator guidance (run diagnostics and avoid deleting legacy artifacts).

[Review Change Stack](https://app.coderabbit.ai/change-stack/gsd-build/gsd-2/pull/5765)

### CODE REVIEW

**Comment by jeremymcs** [comment]
🤖 Dispatched `codex` with the following prompt:

<details>
<summary>Agent prompt (click to expand)</summary>

```text
You are acting as an autonomous PR babysitter for gsd-build/gsd-2 PR #5765.
PR URL: https://github.com/gsd-build/gsd-2/pull/5765
Base repository: gsd-build/gsd-2
Head repository: jeremymcs/gsd-2
Head branch: feat/adr-013-memory-consolidation-scanner
Head remote: fork-jeremymcs
You are running inside an isolated app-owned worktree under ~/.oh-my-pr.
Make only targeted changes that resolve the approved tasks.
Do not wait for user input, confirmation, or approval at any point.
Do not rewrite unrelated files.
Use the available git tooling for inspection and verification only.
If dependencies are missing, install them using the repository's lockfile/package manager as needed inside this isolated worktree.
Leave file edits uncommitted; the babysitter app will handle Git finalization after your run.
GitHub follow-up replies and review-thread resolution will be handled by the babysitter after your run.
If a task is invalid after inspection, explain it in your final response and include the exact audit token.

Approved review-comment tasks:
None

Approved status-check tasks:
None

Approved documentation tasks:
Documentation updates are required for this PR.
Assessment summary: Adds a user/operator-visible memory consolidation preflight warning on startup and references /doctor details, so repository/operator docs or ADR-013 migration docs should explain the scanner, warning meaning, and expected remediation before cutover.
Update the appropriate repository documentation for these changes.
Choose the right docs files for this repository (for example README, docs pages, API/config/operator docs).
If, after inspection, the repository documentation is already accurate or there is no appropriate docs target, leave docs unchanged and report that using the docs summary block with `no_change`.

When done:
1) Run the relevant verification for your changes.
2) Leave any changed files in the worktree for the babysitter app to finalize.
3) For each feedback item you addressed or were blocked on, emit a summary block in the following format:
   FEEDBACK_SUMMARY_START <auditToken>
   <A concise 1-2 sentence summary of what you did or why you were blocked>
   FEEDBACK_SUMMARY_END
   Include one block per audit token. These summaries will be posted as follow-up comments on the PR.
4) If documentation tasks were assigned, emit exactly one docs summary block in the following format:
   DOCS_SUMMARY_START <changed|no_change>
   <A concise 1-2 sentence summary of the docs you updated, or why no docs changes were necessary after inspection>
   DOCS_SUMMARY_END
```

</details>

**Comment by jeremymcs** [comment]
## ⚠️ CI Alert

The babysitter pushed changes (commit `5fc5f86`), but CI/CD checks are still failing:

- **build**: Check run failure ([details](https://github.com/gsd-build/gsd-2/actions/runs/25646458585/job/75276222881))

Manual investigation may be required.

File: src/resources/extensions/gsd/memory-consolidation-scanner.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    logWarning("memory-consolidation", report.summary);
+    return report;
+  } catch (e) {
+    logWarning(
+      "memory-consolidation",
+      `scanner failed: ${(e as Error).message}`,
+    );
+    return null;
+  }
+}
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/memory-consolidation-scanner.ts:277]: ⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `codex` is working on the fix...
✅ **Agent completed** — verifying changes...
🎉 **Resolved** — addressed in commit `af13b72`.

**Comment by jeremymcs** [comment]
⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `codex` is working on the fix...
✅ **Agent completed** — verifying changes...
🎉 **Resolved** — addressed in commit `af13b72`.

**Comment by jeremymcs** [comment]
🤖 Dispatched `codex` with the following prompt:

<details>
<summary>Agent prompt (click to expand)</summary>

````text
You are acting as an autonomous PR babysitter for gsd-build/gsd-2 PR #5765.
PR URL: https://github.com/gsd-build/gsd-2/pull/5765
Base repository: gsd-build/gsd-2
Head repository: jeremymcs/gsd-2
Head branch: feat/adr-013-memory-consolidation-scanner
Head remote: fork-jeremymcs
You are running inside an isolated app-owned worktree under ~/.oh-my-pr.
Make only targeted changes that resolve the approved tasks.
Do not wait for user input, confirmation, or approval at any point.
Do not rewrite unrelated files.
Use the available git tooling for inspection and verification only.
If dependencies are missing, install them using the repository's lockfile/package manager as needed inside this isolated worktree.
Leave file edits uncommitted; the babysitter app will handle Git finalization after your run.
GitHub follow-up replies and review-thread resolution will be handled by the babysitter after your run.
If a task is invalid after inspection, explain it in your final response and include the exact audit token.

Approved review-comment tasks:
1. [review_comment] coderabbitai[bot]
   file=src/resources/extensions/gsd/memory-consolidation-scanner.ts line=266
   sourceId=3215961566 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5765#discussion_r3215961566
   replyKind=review_thread threadId=PRRT_kwDORj5hb86A82h5 threadResolved=false
   auditToken=codefactory-feedback:gh-review-comment-3215961566
   _⚠️ Potential issue_ | _🟡 Minor_ | _⚡ Quick win_

**Inconsistency between docstring and implementation.**

The AI summary states this function "returns `null` when the scan fails or when there are no gaps," but the implementation at line 255 returns the report when `totalGaps === 0`. The test suite (line 304-316) validates the current behavior (returning a truthy report when clean), so the implementation is correct but the documentation is misleading.




<details>
<summary>📝 Suggested docstring clarification</summary>

```diff
 /**
  * Run the scanner and emit a persistent notification + workflow-logger
  * warning when gaps exist. Best-effort: never throws; a broken scanner
  * cannot block agent startup.
  *
+ * Returns the gap report whether gaps exist or not; returns null only on
+ * scanner errors.
+ *
  * Idempotent at the surface: the notification store applies its own
  * 30-second dedup window keyed on (severity, source, message), so repeated
  * boots with identical gaps produce one notification, not a flood.
  */
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/resources/extensions/gsd/memory-consolidation-scanner.ts` around lines
252 - 266, The docstring for reportConsolidationGaps is inconsistent with the
implementation: update the function's comment to state that it returns a
ConsolidationGapReport (including when totalGaps === 0) and only returns null on
scan failure/exception; reference the reportConsolidationGaps function and its
call to scanConsolidationGaps and the behavior that
appendNotification/logWarning run only when report.totalGaps > 0 so the
docstring matches the tests and current logic.
```

</details>

<!-- fingerprinting:phantom:poseidon:puma -->

<!-- d98c2f50 -->

<!-- This is an auto-generated comment by CodeRabbit -->
2. [review] coderabbitai[bot]
   file=n/a line=n/a
   sourceId=4260326377 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5765#pullrequestreview-4260326377
   replyKind=review threadId=n/a threadResolved=n/a
   auditToken=codefactory-feedback:gh-review-4260326377
   **Actionable comments posted: 2**

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

Inline comments:
In `@src/resources/extensions/gsd/bootstrap/system-context.ts`:
- Around line 168-177: The code calls reportConsolidationGaps(basePath) but
basePath is undefined; replace the undefined variable with process.cwd() so the
import call becomes reportConsolidationGaps(process.cwd()); update the call site
inside the try block where reportConsolidationGaps is invoked and keep the
existing catch using logWarning to report failures.

In `@src/resources/extensions/gsd/memory-consolidation-scanner.ts`:
- Around line 252-266: The docstring for reportConsolidationGaps is inconsistent
with the implementation: update the function's comment to state that it returns
a ConsolidationGapReport (including when totalGaps === 0) and only returns null
on scan failure/exception; reference the reportConsolidationGaps function and
its call to scanConsolidationGaps and the behavior that
appendNotification/logWarning run only when report.totalGaps > 0 so the
docstring matches the tests and current logic.
```

</details>

<details>
<summary>🪄 Autofix (Beta)</summary>

Fix all unresolved CodeRabbit comments on this PR:

- [ ] <!-- {"checkboxId": "4b0d0e0a-96d7-4f10-b296-3a18ea78f0b9"} --> Push a commit to this branch (recommended)
- [ ] <!-- {"checkboxId": "ff5b1114-7d8c-49e6-8ac1-43f82af23a33"} --> Create a new PR with the fixes

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro Plus

**Run ID**: `55e98e90-0b8b-49ba-b648-963bec913917`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between eafe113981bc154b5144e54a5d873fe0f5a45c71 and a3e245a3c59e942eb1cdf1824d060514c7a45b93.

</details>

<details>
<summary>📒 Files selected for processing (3)</summary>

* `src/resources/extensions/gsd/bootstrap/system-context.ts`
* `src/resources/extensions/gsd/memory-consolidation-scanner.ts`
* `src/resources/extensions/gsd/tests/memory-consolidation-scanner.test.ts`

</details>

</details>

<!-- This is an auto-generated comment by CodeRabbit for review status -->

Approved status-check tasks:
None

Approved documentation tasks:
None

When done:
1) Run the relevant verification for your changes.
2) Leave any changed files in the worktree for the babysitter app to finalize.
3) For each feedback item you addressed or were blocked on, emit a summary block in the following format:
   FEEDBACK_SUMMARY_START <auditToken>
   <A concise 1-2 sentence summary of what you did or why you were blocked>
   FEEDBACK_SUMMARY_END
   Include one block per audit token. These summaries will be posted as follow-up comments on the PR.
4) If documentation tasks were assigned, emit exactly one docs summary block in the following format:
   DOCS_SUMMARY_START <changed|no_change>
   <A concise 1-2 sentence summary of the docs you updated, or why no docs changes were necessary after inspection>
   DOCS_SUMMARY_END
````

</details>

File: src/resources/extensions/gsd/memory-consolidation-scanner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/memory-consolidation-scanner.ts:277]: Addressed in commit `af13b72` by the latest babysitter run.

Updated the `reportConsolidationGaps` docstring to state that it returns a `ConsolidationGapReport` even when `totalGaps === 0`, returns `null` only on scanner failure, and emits notification/log warnings only when gaps exist.

**Comment by jeremymcs** [comment]
Addressed in commit `af13b72` by the latest babysitter run.

> Responding to [comment by @coderabbitai[bot]](https://github.com/gsd-build/gsd-2/pull/5765#pullrequestreview-4260326377):
> **Actionable comments posted: 2**

Fixed the undefined `basePath` call by using `process.cwd()` and updated the scanner docstring to match the implementation. Focused scanner tests pass; full extension typecheck is blocked by existing unrelated unresolved workspace package errors.

**Comment by jeremymcs** [comment]
## ⚠️ CI Alert

The babysitter pushed changes (commit `af13b72`), but CI/CD checks are still failing:

- **build**: Check run failure ([details](https://github.com/gsd-build/gsd-2/actions/runs/25646595693/job/75276607175))

Manual investigation may be required.

File: src/resources/extensions/gsd/memory-consolidation-scanner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/memory-consolidation-scanner.ts:277]: ⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `codex` is working on the fix...
✅ **Agent completed** — verifying changes...
🎉 **Resolved** — addressed in commit `b3cb50e`.

**Comment by jeremymcs** [comment]
🤖 Dispatched `codex` with the following prompt:

<details>
<summary>Agent prompt (click to expand)</summary>

````text
You are acting as an autonomous PR babysitter for gsd-build/gsd-2 PR #5765.
PR URL: https://github.com/gsd-build/gsd-2/pull/5765
Base repository: gsd-build/gsd-2
Head repository: jeremymcs/gsd-2
Head branch: feat/adr-013-memory-consolidation-scanner
Head remote: fork-jeremymcs
You are running inside an isolated app-owned worktree under ~/.oh-my-pr.
Make only targeted changes that resolve the approved tasks.
Do not wait for user input, confirmation, or approval at any point.
Do not rewrite unrelated files.
Use the available git tooling for inspection and verification only.
If dependencies are missing, install them using the repository's lockfile/package manager as needed inside this isolated worktree.
Leave file edits uncommitted; the babysitter app will handle Git finalization after your run.
GitHub follow-up replies and review-thread resolution will be handled by the babysitter after your run.
If a task is invalid after inspection, explain it in your final response and include the exact audit token.

Approved review-comment tasks:
1. [review_comment] coderabbitai[bot]
   file=src/resources/extensions/gsd/memory-consolidation-scanner.ts line=270
   sourceId=3215981836 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5765#discussion_r3215981836
   replyKind=review_thread threadId=PRRT_kwDORj5hb86A82h5 threadResolved=true
   auditToken=codefactory-feedback:gh-review-comment-3215981836
   `@jeremymcs`, here's the exact docstring patch to apply — it makes the return-value contract explicit without touching any logic:

```diff
-/**
- * Run the scanner and emit a persistent notification + workflow-logger
- * warning when gaps exist. Best-effort: never throws; a broken scanner
- * cannot block agent startup.
- *
- * Idempotent at the surface: the notification store applies its own
- * 30-second dedup window keyed on (severity, source, message), so repeated
- * boots with identical gaps produce one notification, not a flood.
- */
+/**
+ * Run the scanner and emit a persistent notification + workflow-logger
+ * warning when gaps exist. Best-effort: never throws; a broken scanner
+ * cannot block agent startup.
+ *
+ * Returns the full {`@link` ConsolidationGapReport} regardless of whether gaps
+ * exist (including when `totalGaps === 0`). Returns `null` only when
+ * `scanConsolidationGaps` itself throws. `appendNotification` and
+ * `logWarning` are called only when `report.totalGaps > 0`.
+ *
+ * Idempotent at the surface: the notification store applies its own
+ * 30-second dedup window keyed on (severity, source, message), so repeated
+ * boots with identical gaps produce one notification, not a flood.
+ */
```

No logic changes needed — this is purely documentation to match what the tests and implementation already guarantee.

<!-- This is an auto-generated reply by CodeRabbit -->

Approved status-check tasks:
None

Approved documentation tasks:
None

When done:
1) Run the relevant verification for your changes.
2) Leave any changed files in the worktree for the babysitter app to finalize.
3) For each feedback item you addressed or were blocked on, emit a summary block in the following format:
   FEEDBACK_SUMMARY_START <auditToken>
   <A concise 1-2 sentence summary of what you did or why you were blocked>
   FEEDBACK_SUMMARY_END
   Include one block per audit token. These summaries will be posted as follow-up comments on the PR.
4) If documentation tasks were assigned, emit exactly one docs summary block in the following format:
   DOCS_SUMMARY_START <changed|no_change>
   <A concise 1-2 sentence summary of the docs you updated, or why no docs changes were necessary after inspection>
   DOCS_SUMMARY_END
````

</details>

File: src/resources/extensions/gsd/memory-consolidation-scanner.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/memory-consolidation-scanner.ts:277]: Addressed in commit `b3cb50e` by the latest babysitter run.

Updated the `reportConsolidationGaps` docstring to explicitly state it returns the full `ConsolidationGapReport` for all successful scans, returns `null` only when `scanConsolidationGaps` throws, and only emits notifications/warnings when `report.totalGaps > 0`. Focused scanner tests pass.

**Comment by jeremymcs** [comment]
## ⚠️ CI Alert

The babysitter pushed changes (commit `b3cb50e`), but CI/CD checks are still failing:

- **build**: Check run failure ([details](https://github.com/gsd-build/gsd-2/actions/runs/25646706956/job/75276914714))

Manual investigation may be required.

**Comment by jeremymcs** [comment]
⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `codex` is working on the fix...
✅ **Agent completed** — verifying changes...
🎉 **Resolved** — addressed in commit `5a0f7c4`.

**Comment by jeremymcs** [comment]
🤖 Dispatched `codex` with the following prompt:

<details>
<summary>Agent prompt (click to expand)</summary>

```text
You are acting as an autonomous PR babysitter for gsd-build/gsd-2 PR #5765.
PR URL: https://github.com/gsd-build/gsd-2/pull/5765
Base repository: gsd-build/gsd-2
Head repository: jeremymcs/gsd-2
Head branch: feat/adr-013-memory-consolidation-scanner
Head remote: fork-jeremymcs
You are running inside an isolated app-owned worktree under ~/.oh-my-pr.
Make only targeted changes that resolve the approved tasks.
Do not wait for user input, confirmation, or approval at any point.
Do not rewrite unrelated files.
Use the available git tooling for inspection and verification only.
If dependencies are missing, install them using the repository's lockfile/package manager as needed inside this isolated worktree.
Leave file edits uncommitted; the babysitter app will handle Git finalization after your run.
GitHub follow-up replies and review-thread resolution will be handled by the babysitter after your run.
If a task is invalid after inspection, explain it in your final response and include the exact audit token.

Approved review-comment tasks:
1. [general_comment] jeremymcs
   file=n/a line=n/a
   sourceId=4417072665 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5765#issuecomment-4417072665
   replyKind=general_comment threadId=n/a threadResolved=n/a
   auditToken=codefactory-feedback:gh-issue-comment-4417072665
   ## ⚠️ CI Alert

The babysitter pushed changes (commit `b3cb50e`), but CI/CD checks are still failing:

- **build**: Check run failure ([details](https://github.com/gsd-build/gsd-2/actions/runs/25646706956/job/75276914714))

Manual investigation may be required.

Approved status-check tasks:
None

Approved documentation tasks:
None

When done:
1) Run the relevant verification for your changes.
2) Leave any changed files in the worktree for the babysitter app to finalize.
3) For each feedback item you addressed or were blocked on, emit a summary block in the following format:
   FEEDBACK_SUMMARY_START <auditToken>
   <A concise 1-2 sentence summary of what you did or why you were blocked>
   FEEDBACK_SUMMARY_END
   Include one block per audit token. These summaries will be posted as follow-up comments on the PR.
4) If documentation tasks were assigned, emit exactly one docs summary block in the following format:
   DOCS_SUMMARY_START <changed|no_change>
   <A concise 1-2 sentence summary of the docs you updated, or why no docs changes were necessary after inspection>
   DOCS_SUMMARY_END
```

</details>

**Comment by jeremymcs** [comment]
Addressed in commit `5a0f7c4` by the latest babysitter run.

> Responding to [comment by @jeremymcs](https://github.com/gsd-build/gsd-2/pull/5765#issuecomment-4417072665):
> ## ⚠️ CI Alert

Fixed the reproduced TypeScript build failure by safely mapping raw DB rows in the memory consolidation scanner and registering `memory-consolidation` as a valid workflow log component. Verified with `npm run build:core`, `npm run typecheck:extensions`, and the scanner unit test; local web-host verification was blocked by restricted npm registry access.
